src/blockchain_db/sqlite/db_sqlite.cpp

```diff
@@ -0,0 +1,666 @@
+  return success;
+}
+
+bool BlockchainSQLite::pop_block(const cryptonote::block& block, const service_nodes::service_node_list::state_t& service_nodes_state)
+{
+  auto block_height = get_block_height(block);
+
+  LOG_PRINT_L3("BlockchainDB_SQLITE::" << __func__ << " called on height: " << block_height);
+  if (height < block_height) {
+    MDEBUG("Block above batching DB height skipping pop");
+    return true;
+  }
+  if (cryptonote::get_block_height(block) != height)
+  {
+    MERROR("Block height out of sync with batching database");
+    return false;
+  }
+
+  const auto& conf = get_config(m_nettype);
+  auto hf_version = block.major_version;
+  if (hf_version < cryptonote::network_version_19)
+    return decrement_height();
+
+
+  bool success = false;
+  try
+  {
+    SQLite::Transaction transaction{db, SQLite::TransactionBehavior::IMMEDIATE};
+
+
+    // Step 1: Remove the block producers txn fees
+    uint64_t service_node_reward = cryptonote::service_node_reward_formula(0, block.major_version);
+    if (block.reward - service_node_reward > 0)
+    {
+      if (block.service_node_winner_key && crypto_core_ed25519_is_valid_point(reinterpret_cast<const unsigned char *>(block.service_node_winner_key.data)))
+      {
+        auto service_node_winner = service_nodes_state.service_nodes_infos.find(block.service_node_winner_key);
+        if (service_node_winner != service_nodes_state.service_nodes_infos.end())
+        {
+          std::vector<cryptonote::batch_sn_payment> block_producer_fee_payments = calculate_rewards(block.major_version, block.reward - service_node_reward, *service_node_winner->second);
+          // Takes the block producer and adds its contributors to the batching database for the transaction fees
+          success = subtract_sn_payments(block_producer_fee_payments);
+          if (!success)
+            return false;
+          }
+        }
+    }
+
+    // Step 2: Iterate over the whole service node list and subtract each node 1/service_node_list fraction
+    const auto payable_service_nodes = service_nodes_state.payable_service_nodes_infos(block_height, m_nettype);
+    size_t total_service_nodes_payable = payable_service_nodes.size();
+    for (const auto& [node_pubkey, node_info]: payable_service_nodes)
+    {
+      std::vector<cryptonote::batch_sn_payment> node_contributors;
+      auto payable_service_node = service_nodes_state.service_nodes_infos.find(node_pubkey);
+      if (payable_service_node == service_nodes_state.service_nodes_infos.end())
+        continue;
+      std::vector<cryptonote::batch_sn_payment> node_rewards = calculate_rewards(block.major_version, service_node_reward/total_service_nodes_payable, *payable_service_node->second);
+      // Takes the node and adds its contributors to the batching database
+      success = subtract_sn_payments(node_rewards);
+      if (!success)
+        return false;
+    }
+    // Step 3: Remove Governance reward
+    if (m_nettype != cryptonote::FAKECHAIN)
```

Comment:
Why this?  Does fakechain not have a governance reward?

Reply:
Yeah the tests dont ever pay a governance reward. They just want to add blocks with whatever they want and not worry about it.